### ISSUE TITLE
supporting schema evolution

### ISSUE THREAD

**Issue by alanpaulkwan** [duckdb/duckdb#2877]
For reasons not worth discussing, I have data which 

- same field evolves over time to be slightly different (FLOAT vs DOUBLE vs BIGINT vs INT etc)
- new field

DuckDB will first check the formats and error out if the schema are incomptabile across multiple Parquet files.

Desired behavior:

- Evaluate failure upon casting, so if I say cast(field_with_evolving_type as INTEGER) it will attempt the cast before declaring failure
- Impute missing values if I ask for a column which does not exist in all files, but at least in one

I understand that this may not be desirable default behavior but right now the only option is to rewrite the underlying files to have the same schema or something along those lines.

**Comment by hannesmuehleisen** [comment]
There are two things to consider here:

1. Types, we get the DuckDB types from looking at the first file's metadata, and we cannot look at all files' metadata since this happens during the bind phase which is in the critical single-threaded path for query planning. Hence, we would need to try to cast all the columns from other files to the types found in the first file. Is this acceptable to you?
2. Missing/added columns between files. This one is trickier, as again we need to decide the number, name and types of columns during the bind. If one of the other files is missing a column, that would be doable. If one of the subsequent files has additional columns, that would be impossible. Deal-breaker?

**Comment by alanpaulkwan** [comment]
This sounds OK, but most schema evolution is probably because you add over time and thus big-endian names would make the first file the worst file, probably. Maybe there's a way to specify the template file? This solves the problem with both. So if for example the last file was the superset file, we could do something like

parquet_scan('./*.parquet',  format='2021-12-31.parquet' , fill=TRUE,cast=TRUE)

**Comment by hannesmuehleisen** [comment]
You could also manually specify the parquet files in the order they are to be read with a list parameter, having the broadest file first in that list.

**Comment by alanpaulkwan** [comment]
Do you mean manually rewriting the broadest file to come first alphabetically, or is there some syntax to supply them with a list? If the former, that would not work in the S3 world since you'd have to rewrite files and so forth incurring costs.

**Comment by Alex-Monahan** [comment]
There is a way to use a list as an input to parquet_scan!
[Parquet Docs and Examples](https://duckdb.org/docs/data/parquet)
```sql
-- use list parameter to read 3 parquet files and treat them as a single table 
SELECT * FROM parquet_scan(['file1.parquet', 'file2.parquet', 'file3.parquet']);
-- use a list of globs to read all parquet files from 2 specific folders
SELECT * FROM parquet_scan(['folder1/*.parquet','folder2/*.parquet']);
```

**Comment by alanpaulkwan** [comment]
DuckDB is amazing!

**Comment by arjenpdevries** [comment]
DuckDB is even more amazing!!

Checkout this example code to turn a table, into which you would insert the filenames of your parquet filenames as they are being created, into a list:

```
--
-- CTE to convert table to list
--
-- Credits:
-- https://www.sqlmatters.com/Articles/Converting%20row%20values%20in%20a%20table%20to%20a%20single%20concatenated%20string.aspx

create table vs (v varchar);
insert into vs (values ('f1.parquet'), ('f2.parquet'), ('f3.parquet'));

with recursive vl as
  (
     select   rowid, [v] as l
     from     vs
     where    rowid=0
   union all
     select   vs.rowid, list_append(vl.l, vs.v)
     from     vl
     join     vs
     on       vs.rowid = vl.rowid + 1
   )
   SELECT   l
   FROM     vl
   WHERE    rowid = (SELECT MAX(rowid) FROM vs);
```

**Comment by Mytherin** [comment]
Implemented in  #3238

**Comment by dberardo-com** [comment]
dear all, 
i am struggling to find a complete list of the arguments / options accepted by the parquet_scan operator.

is there any reference i have missed in the doc ?
